File: site/public/sites/vqx/crypto/vqx_crypto.py
"""Optional VQX envelopes. Encoding is not encryption."""

from __future__ import annotations

import hashlib
import hmac
import os


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def hmac_sha256(key: bytes, data: bytes) -> bytes:
    return hmac.new(key, data, hashlib.sha256).digest()


def encrypt_aes256_gcm(key: bytes, plaintext: bytes, aad: bytes = b""):
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    if len(key) != 32:
        raise ValueError("AES-256-GCM key must be 32 bytes")
    iv = os.urandom(12)
    aes = AESGCM(key)
    packed = aes.encrypt(iv, plaintext, aad if aad else None)
    return {"iv": iv, "ciphertext": packed[:-16], "tag": packed[-16:]}


def decrypt_aes256_gcm(key: bytes, iv: bytes, ciphertext: bytes, tag: bytes, aad: bytes = b""):
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    aes = AESGCM(key)
    return aes.decrypt(iv, ciphertext + tag, aad if aad else None)


def generate_ed25519():
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

    private = Ed25519PrivateKey.generate()
    return private, private.public_key()


def sign_ed25519(private_key, data: bytes) -> bytes:
    return private_key.sign(data)


def verify_ed25519(public_key, data: bytes, signature: bytes) -> bool:
    from cryptography.exceptions import InvalidSignature

    try:
        public_key.verify(signature, data)
        return True
    except InvalidSignature:
        return False

File: tools/vqx/crypto/vqx-crypto.mjs
/** Optional VQX envelopes. Encoding is not encryption. */
import { createHmac, createCipheriv, createDecipheriv, randomBytes, createHash, generateKeyPairSync, sign as nodeSign, verify as nodeVerify } from "node:crypto";

export function sha256(buf) {
  return createHash("sha256").update(buf).digest("hex");
}

export function hmacSha256(key, buf) {
  return createHmac("sha256", key).update(buf).digest();
}

export function encryptAes256Gcm(key, plaintext, aad = Buffer.alloc(0)) {
  if (key.length !== 32) throw new Error("AES-256-GCM key must be 32 bytes");
  const iv = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", key, iv);
  if (aad.length) cipher.setAAD(aad);
  const ct = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();
  return { iv, ciphertext: ct, tag };
}

export function decryptAes256Gcm(key, iv, ciphertext, tag, aad = Buffer.alloc(0)) {
  const decipher = createDecipheriv("aes-256-gcm", key, iv);
  if (aad.length) decipher.setAAD(aad);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

export function generateEd25519() {
  return generateKeyPairSync("ed25519");
}

export function signEd25519(privateKey, data) {
  return nodeSign(null, data, privateKey);
}

export function verifyEd25519(publicKey, data, signature) {
  return nodeVerify(null, data, publicKey, signature);
}
